Daemon fixes

pmorch opened the pull request:
There were a number of problems with the daemon, that I've taken the liberty to fix. I hope you don't find any of them objectionable.

andreas-marschke commented:
@pmorch Hey Peter!

Thanks for the changes. Can you elaborate here in the PR description on the Changes you've applied?  Your commit messages look wonderful and detailed, for documentation purposes here in the Github repo I think it'd be awesome to also add them here to the PR.

Thanks!

pmorch commented:
I think you'd like me to add the commit messages inline. Correct? Something like this:

```
commit 0e1e8a753ec51736a8c81a0de7d6bdc87613ef12
Author: Peter Valdemar Mørch <peter@morch.com>
Date:   Wed Mar 7 16:18:29 2018 +0100

    Remove syntax error from non-used documentation code

commit 4298464c429a047fa56882e8120224cc936478e1
Author: Peter Valdemar Mørch <peter@morch.com>
Date:   Wed Mar 7 16:21:53 2018 +0100

    Nagios::NRPE::Daemon: Ran perltidy on example code

commit ebb4cebf6a106ae5dcf6b5f497c8ed5716c9a8c1
Author: Peter Valdemar Mørch <peter@morch.com>
Date:   Wed Mar 7 16:23:21 2018 +0100

    Nagios::NRPE::Daemon: Doc example - actually start the daemon that gets created

commit b354865e3271c21c5fb1c7470b4c56d172efca99
Author: Peter Valdemar Mørch <peter@morch.com>
Date:   Wed Mar 7 16:24:33 2018 +0100

    Nagios::NRPE::Packet: Set result_code to 2324 only if it is not defined
    
    Before result_code got set if it was false. That had the problem that
    if the result_code explicity got set to STATUS_OK (which is 0 and
    therefore perl false) it got set the 2324 which is a nagios error.

commit 10021b0da93cad65db52d8d30b9c3a2fd9356eb8
Author: Peter Valdemar Mørch <peter@morch.com>
Date:   Wed Mar 7 17:02:57 2018 +0100

    Nagios::NRPE::Daemon: Fix multiple problems with callbacks
    
    This fixes both the example callbacks and the callback in
    bin/nrpe-server.pl
    
    * Allow callback to specify status code
      Allow specifying the status code (ok, warning, critical, etc.) as
      well as output message. Before the callback could only specify the
      output message, but nagios uses both
    * The substitution of $ARGN$ strings simply didn't work
    * The entire STDERR+STDOUT output from the child would be sent back as
      the message. This is inconsistent with Nagios that only considers
      STDOUT and ignores STDERR
    * If the daemon is called with a $check that wasn't configured, it
      wouldn't return anything and check_nrpe would say:
        CHECK_NRPE: Received 0 bytes from daemon.  Check the remote server logs for error messages.
        exit code: 3
      Now it says:
        sprintf("No such check: '%s'", $check_name)
        exit code: 3
    
    nrpe-server.pl and the examples still don't handle exit codes more
    elaborately than checking for $success, but that is due to a
    limitation of IPC::Cmd::run, and could be fixed by using IPC::Run, but
    that introduces a new dependency and was more hassle than I was
    prepared to deal with.

commit 334804a8bc4c62af4ddd3517d21d4ce5c2497a84
Author: Peter Valdemar Mørch <peter@morch.com>
Date:   Wed Mar 7 17:49:40 2018 +0100

    Remove misleading shebang from .pm files and leading =pod-s too

commit 039d2f1d798b830bf7201001d4b03c28b3d8d6c9
Author: Peter Valdemar Mørch <peter@morch.com>
Date:   Wed Mar 7 18:01:19 2018 +0100

    Removed misleading mention of threads in bin/nrpe-server.pl
    
    bin/nrpe-server.pl had:
    
        use threads;
        <snip>
        threads->new( $daemon->start() );
    
    On the surface this looks like this makes the daemon magically be
    multithreaded and handle multiple requests in parallel.
    
    Sadly that is not the case, as running to requests in parallel against
    the $daemon will demonstrate. It is single-threaded.
    
    $daemon->start() never returns, so threads->new() is never called. The
    threads code does absolutely nothing.
    
    So it is no-op and misleading. In addition, perldoc threads says:
    
        The "interpreter-based threads" provided by Perl are not the fast,
        lightweight system for multitasking that one might expect or hope for.
        Threads are implemented in a way that make them easy to misuse. Few people
        know how to use them correctly or will be able to provide help.
    
        The use of interpreter-based threads in perl is officially discouraged.
    
    So I simply removed the misleading no-op threads code and replaced:
    
        use threads;
        <snip>
        threads->new( $daemon->start() );
    
    with:
        $daemon->start();
```

andreas-marschke commented:
Less the messages themselves but a summary based on the wonderful commit messages you've written.

pmorch commented:
Ok, a summary:

I tried to create a daemon based on the perldoc and noticed many things didn't work, and that some things were just weird:

First of all, the example in the perldoc was clearly not working - it had syntax errors. The perldoc examples also benefited from perltidy to make them easier to read.

Next, I noticed that the daemon was created, but not actually started.

After I got it started, I noticed that whatever I did, `/usr/lib/nagios/plugins/check_nrpe $args ; echo $?` always showed 20 and the callback didn't allow to specify an exitcode or status. Once I really looked at the callbacks, there were other issues. See the commit message for 10021b0da93cad65db52d8d30b9c3a2fd9356eb8.

I then discovered bin/nrpe-server.pl and was confused about it mentioning `threads`, but it wasn't threaded, really. The threading was non-existent so I removed it.

Cosmetically I removed the `#!/usr/bin/perl -w` shebang lines from the .pm files, because .pm files aren't executable anyway.
